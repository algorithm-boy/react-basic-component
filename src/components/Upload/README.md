# Upload (上传组件)

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| action(**required**) | 上传文件的URL或一个回调函数 (Uploading URL or Callback) | string \| (file) => Promise | - |
| className(optional) | Upload的类名（The class property for Upload) | string | '' |
| style(optional) | Upload的行内样式 (The inline style for Upload) | object | {} |
| id(optional) | The input id(input的id) | string | - |
| prefixCls(optional) | 类名前缀 (The prefix of class) | boolean | 'hfdc-upload' |
| name(optional) | 上传文件的文件名 (The name of uploading file(s)) | string \| array | - |
| accept(optional) | 上传文件的类型, 预设值有**all、image、audio、video** (File type that can be accepted) | string | '' |
| size(optional) | 上传文件的最大值，单位MB (The max size of uploading file, unit is MB) | number | Infinity |
| multiple(optional) | 是否允许多文件上传 (Whether allow multiple files upload) | boolean | false |
| minCount(optional) | 上传文件的最小个数, 默认为1 (Minimum number of uploaded files) | number | 1 |
| maxCount(optional) | 上传文件的最大个数, `mutiple`为`false`时，不生效 (Maximum number of uploaded files, it didn't work when `multiple` is `false`) | number | Infinity |
| uploadModel(optional) | 上传模式，默认并发多个上传请求，可选`concatenation`为串行上传 (Uploading Model, default is concurrent) | concurrent \| concatenation | concurrent |
| concatenation(optional) | 上传模式为`concatenation`串行上传时，串行的个数 (Number of series when Upload Model is `concatenation`) | number | 1 |
| disabled(optional) | 禁用上传 (disable uploading) | boolean | false |
| withCredentials(optional) | 请求时是否带上cookie (ajax upload with cookie sent) | boolean | false |
| beforeUpload(optional) | 开始上传之前的回调函数，返回`false`会终止上传 (A callback function will be called when before uploading step. The uploading will be stop when the function return `false`) | (isRequesting: boolean, files: ChangeData, e?: ChangeEvent) => boolean; | - |
| onChange(optional) | 上传状态改变的回调函数 (A callback function will be called when uploading state had been changed) | (files: ChangeData, e: ChangeEvent) => any | - |
| onProgress(optional) | 上传进度的回调函数 (A callback function will be called when uploading) | (e: ProgressEv) => any | - |
| onSuccess(optional) | 上传成功的回调函数 (A callback function will be called when have been uploaded) | (res: any) => any | - |
| onFinish(optional) | 所有文件均执行上传操作后的回调函数 (A callback function will be called when all files has been completed upload procedures) | () => any | - |
| onException(optional) | 上传过程中出现错误的回调函数，type可能是`size、count、accept、api`分别代表不匹配设定的size、maxCount、accept，或者是上传的api报错 (A callback function will be called when the uploading process occur any error) | (type: ErrType, e?: ErrorEvent \| Err \| ProgressEvent) => any | - |

---

## Types
```typescript
export interface ProgressEv extends ProgressEvent {
  percent: number;
}
export interface Err extends Error{
  status: number;
  method: string;
  url: string;
}
interface ChangeFile {
  label: string;
  file: File;
}
type ChangeData = ChangeFile[];
type ActionFn = (files: ChangeData) => Promise<any>;
type Preset = 'all' | 'image' | 'audio' | 'video';
type ErrType = 'size' | 'count' | 'accept' | 'api';
type UploadModel = 'concurrent' | 'concatenation';
type DomProps = 'id' | 'style' | 'multiple' | 'disabled';
```

---

## Example

```javascript
<button style={{ position: 'relative' }}>
  Upload
  <Upload action='http://hupu.com' />
</button>
```

