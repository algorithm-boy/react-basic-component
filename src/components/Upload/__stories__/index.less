@prefix: hfdc-demo;

.mixin-add (@size: 14px, @bold: 2px, @color:#fc5832) {
  width: @size;
  height: @size;
  position: absolute;
  &::before,
  &::after{
    content: "";
    position: absolute;
    height: 0.8 * @size;
    width: @bold;
    top: 0.1 * @size;
    right: (@size - @bold) / 2;
    background-color: @color;
    border-radius: 0.5 * @bold;
  }
  &::before{
    transform: rotate(180deg);
  }
  &::after{
    transform: rotate(-90deg);
  }
}

.@{prefix}-upload{
  position: relative;
  margin: 50px auto;
  border: 1px dashed gray;
  width: 200px;
  height: 200px;
  &-btn{
    display: block;
    position: relative;
    width: 200px;
    height: 50px;
    margin: 0 auto;
    font-size: 20px;
    color: #fc5832;
  }
  &-cross{
    .mixin-add(60px, 3px);
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
  &-article{
    margin: 0 auto;
    text-align: center;
    font-size: 22px;
    color: crimson;
  }
}