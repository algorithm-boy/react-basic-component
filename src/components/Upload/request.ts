import { isType } from 'peeler-js';

export interface Headers {
  [propName: string]: string;
}

export interface Data {
  [propName: string]: any;
}

export interface ProgressEv extends ProgressEvent {
  percent: number;
}


export interface Option {
  action: string;
  filename?: string;
  file?: File;
  data?: Data[];
  withCredentials?: boolean;
  headers?: Headers;
  onProgress?: (e: ProgressEv) => any;
  onError?: (e: ProgressEvent | Err, res?: any) => any;
  onSuccess?: (res: any) => any;
}

export interface Err extends Error{
  status: number;
  method: string;
  url: string;
}

function getError(option: Option, xhr: XMLHttpRequest): Err {
  const msg = 'cannot post ' + option.action + ' ' + xhr.status + '\'';
  const err = new Error(msg);
  const status = xhr.status;
  const method = 'post';
  const url = option.action;
  return { ...err, status, method, url };
}

function getBody(xhr: XMLHttpRequest) {
  const text = xhr.responseText || xhr.response;
  if (!text) {
    return text;
  }

  try {
    return JSON.parse(text);
  } catch (e) {
    return text;
  }
}

function upload (option: Option) {
  if (typeof XMLHttpRequest === 'undefined') {
    return Promise.reject('Your Browser is out of date');
  }

  return new Promise((resolve, reject) => {
    const { onProgress, onError, onSuccess, filename, file } = option;
    const xhr = new XMLHttpRequest();
    if (xhr.upload) {
      xhr.upload.onprogress = function progress(e) {
        let percent = 0;
        if (e.total > 0) {
          percent = e.loaded / e.total * 100;
        }
        onProgress && onProgress({ ...e, percent });
      };
    }

    const formData = new FormData();

    if (option.data) {
      const data = option.data;
      const len = data.length;
      for (let i = 0; i < len; i++) {
        const item = data[i];
        const key = Object.keys(item)[0];
        const val = item[key];
        formData.append(key, val);
      }
    };

    if (filename && file) {
      formData.append(filename, file);
    }

    xhr.onerror = function error(e) {
      onError ? reject(onError(e)) : reject(e);
    };

    xhr.onload = function onload() {
      if (xhr.status !== 200) {
        const e = getError(option, xhr);
        const res = getBody(xhr);
        return onError ? reject(onError(e, res)) : reject({ e, res });
      };

      onSuccess ? resolve(onSuccess(getBody(xhr))) : resolve(getBody(xhr));
    };

    xhr.open('post', option.action, true);

    // Has to be after `.open()`. See https://github.com/enyo/dropzone/issues/179
    if (option.withCredentials && 'withCredentials' in xhr) {
      xhr.withCredentials = true;
    };

    xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
    const headers = option.headers || {};
    const headerKeys = Object.keys(headers);
    const len = headerKeys.length;
    for (let i = 0; i < len; i++) {
      const key = headerKeys[i];
      if (isType('string')(headers[key]) && headers.hasOwnProperty(key)) {
        const content = headers[key];
        xhr.setRequestHeader(key, content);
      };
    };

    xhr.send(formData);
  });
};

export default upload;