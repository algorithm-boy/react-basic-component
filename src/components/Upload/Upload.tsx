import React, { PureComponent, ChangeEvent } from 'react';
import { Style, GeneralProps } from '../global.d';
import pick from 'lodash/pick';
import { getTs, getUA, isType } from 'peeler-js';
import classnames from 'classnames';
import request, { ProgressEv, Err } from './request';
import { createPropsGetter, logger } from '../../utils';

const preset_accept = {
  all: '',
  image: 'image/*',
  audio: 'audio/*',
  video: 'video/*'
};

enum upload_model {
  concurrent,
  concatenation
}

interface ChangeFile {
  label: string;
  file: File;
}

type ChangeData = ChangeFile[];
type ActionFn = (files: ChangeData) => Promise<any>;
type Preset = 'all' | 'image' | 'audio' | 'video';
type ErrType = 'size' | 'count' | 'accept' | 'api';
type UploadModel = 'concurrent' | 'concatenation';
type DomProps = 'id' | 'style' | 'multiple' | 'disabled';

export interface UploadProps extends GeneralProps {
  action: string | ActionFn;
  accept?: Preset | string;
  prefixCls?: string;
  id?: string;
  name?: string;
  size?: number;
  multiple?: boolean;
  minCount?: number;
  maxCount?: number;
  withCredentials?: boolean;
  disabled?: boolean;
  uploadModel?: UploadModel;
  concatenation?: number;
  beforeUpload?: (isRequesting: boolean, files: ChangeData, e?: ChangeEvent) => boolean | void;
  onChange?: (files: ChangeData, e: ChangeEvent) => any;
  onProgress?: (e: ProgressEv) => any;
  onSuccess?: (res: any) => any;
  onFinish?: () => any;
  onException?: (type: ErrType, e?: ErrorEvent | Err | ProgressEvent) => any;
};

export interface UploadStates {
  isMounted: boolean;
  isIE: boolean;
};

const defaultProps = {
  style: {} as any as Style,
  className: '',
  accept: preset_accept.all as Preset | string,
  prefixCls: 'hfdc-upload',
  size: Infinity,
  multiple: false,
  minCount: 0,
  maxCount: Infinity,
  withCredentials: false,
  disabled: false,
  uploadModel: 'concurrent' as UploadModel,
  concatenation: 1
};

type DefaultProps = Readonly<typeof defaultProps>;

const getProps = createPropsGetter(defaultProps);

class Upload extends PureComponent<UploadProps & Partial<DefaultProps>, UploadStates> {
  static defaultProps = defaultProps;

  private file: React.RefObject<HTMLInputElement>;
  private fileData: ChangeData;
  private isRequesting: boolean;

  public constructor(props: UploadProps) {
    super(props);
    this.state = {
      isMounted: false, // can't use upload until the component mounted
      isIE: false
    };

    this.fileData = [];
    this.isRequesting = false;
    this.file = React.createRef();
    this.handleChange = this.handleChange.bind(this);

    // logic principle checking
    const { minCount, maxCount } = props;
    if (minCount && maxCount && minCount > maxCount) {
      logger.logErr(`Upload - minCount(${minCount}) could not bigger than maxCount(${maxCount})`);
      throw new Error(`minCount(${minCount}) could not bigger than maxCount(${maxCount})`);
    }
  }

  public componentDidMount() {
    const { isIE, isEdge } = getUA(navigator.userAgent);
    this.setState({
      isMounted: true,
      isIE: isIE || isEdge
    });
  }

  public componentWillUnmount() {
    this._clearInput(); // clear upload object
  }

  /**
   * get file size restrict
   * @returns {number}
   */
  private getSize(): number {
    const { size } = getProps(this.props);

    return size * (1024 * 1024);
  }

  /**
   * get acceptable upload file type
   * @returns {string}
   */
  private getAccept(): string {
    const { accept } = getProps(this.props);

    return preset_accept[accept as Preset] || accept;
  }

  /**
   * get upload model
   * @returns {UploadModel}
   */
  private getModel(): 0 | 1 {
    const { uploadModel } = getProps(this.props);

    return upload_model[uploadModel] || upload_model.concurrent;
  }

  private _clearInput() {
    this.file.current && (this.file.current.value = '');
  }

  private _handleAccept(type: string) {
    const accept = this.getAccept();
    if (!accept || accept === 'all' || accept === '*') return true;

    const acceptArr = accept.toLowerCase().split(',');
    const acceptFilter = /\//g;
    const file_type_up = type.toLowerCase().split('/')[0];
    const file_type_sub = type.toLowerCase().split('/')[1];
    const len = acceptArr.length;

    for (var k = 0; k < len; k++) {
      let accept_type_up, accept_type_sub;
      if (acceptFilter.test(acceptArr[k])) {
        accept_type_up = acceptArr[k].split('/')[0];
        accept_type_sub = acceptArr[k].split('/')[1];
      } else {
        accept_type_up = acceptArr[k].split('.')[0];
        accept_type_sub = acceptArr[k].split('.')[1];
      };

      if (accept_type_up === file_type_up) {
        // upper type must be equal
        if (accept_type_sub === '*') {
          // some kind of type accept *
          break;
        } else {
          const regExp = new RegExp(accept_type_sub, 'g');
          if (accept_type_sub === file_type_sub || regExp.test(file_type_sub)) {
            break;
          };
        }
      }
    };

    if (k === acceptArr.length) return false;

    return true;
  }

  private async _handleRequest(e?: ChangeEvent): Promise<void> {
    'use strict';
    const { action, beforeUpload, onProgress, onSuccess, onException, onFinish, name, withCredentials } = getProps(this.props);

    const handleApiErr = (error: ProgressEvent | Err) => {
      this.isRequesting = false;
      logger.logErr(`Upload - The Upload Request occur error - ${JSON.stringify(error).substr(0, 200)}……`);
      onException && onException('api', error);
    };

    const handleApiSuc = (res: any) => {
      this.isRequesting = false;
      onSuccess && onSuccess(res);
    };

    if (!action) {
      logger.logErr('Upload - Please pass correct action');
      return;
    }

    const beforeRes = beforeUpload && beforeUpload(this.isRequesting, this.fileData, e);
    if (beforeRes === false) return;

    const model = this.getModel();
    if (model === 1) {
      // concatenation upload
      let { concatenation } = getProps(this.props);
      if (concatenation < 1) {
        concatenation = 1;
        logger.logWarn('Upload - concatenation must >= 1');
      };

      const uploadData = this.fileData.slice(0, concatenation);
      const len = uploadData.length;
      if (len > 0) {
        this.isRequesting = true; // begin upload and can't continue call upload api before completed this mission
        const { label, file } = uploadData[0];
        const filename = name || file.name || label;

        try {
          if (isType('function')(action)) {
            await action(uploadData)
              .catch(err => handleApiErr(err));
          } else if (isType('string')(action)) {
            if (len > 1) {
              const data = [];
              for (let i = 0; i < len; i++) {
                const item = uploadData[i];
                const { label, file } = item;
                const filename = name || file.name || label;
                data.push({ [filename]: file });
              }
  
              await request({
                action,
                withCredentials,
                data,
                onProgress
              })
                .then(res => handleApiSuc(res))
                .catch(err => handleApiErr(err));
            } else {
              await request({
                action,
                withCredentials,
                file,
                filename,
                onProgress
              })
                .then(res => handleApiSuc(res))
                .catch(err => handleApiErr(err));
            }
          }

          // delete already had been uploaded file
          this.fileData.splice(0, concatenation);
          return this._handleRequest();
        } catch (err) {
          this.isRequesting = false;
          logger.logErr(`Upload - ${JSON.stringify(err)}`);
        }
      } else {
        // all data completed upload
        this.isRequesting = false;
        onFinish && onFinish();
      }
    } else {
      // concurrent upload
      const len = this.fileData.length;
      if (len > 0) {
        this.isRequesting = true;
        try {
          if (isType('function')(action)) {
            await action(this.fileData)
              .catch(err => handleApiErr(err));
          } else if (isType('string')(action)) {
            const data = [];
            for (let i = 0; i < len; i++) {
              const item = this.fileData[i];
              const { label, file } = item;
              const filename = name || file.name || label;
              data.push({ [filename]: file });
            }

            await request({
              action,
              withCredentials,
              data,
              onProgress
            })
              .then(res => handleApiSuc(res))
              .then(() => onFinish && onFinish())
              .catch(err => handleApiErr(err));
          }
        } catch (err) {
          this.isRequesting = false;
          logger.logErr(`Upload - ${JSON.stringify(err)}`);
        } finally {
          // clear fileData
          this.fileData.splice(0);
        }
      } else {
        // all data completed upload
        this.isRequesting = false;
        onFinish && onFinish();
      }
    }
  }

  public handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    'use strict';
    const { isMounted } = this.state;
    const { onChange, onException, multiple, minCount, maxCount, disabled } = getProps(this.props);

    if (!isMounted || disabled) return false;

    const files = e && e.target.files;
    const len = files ? files.length : 0;
    const changeData = [];
    if (files) {
      // there is files upload
      const size = this.getSize();

      if ((!multiple && len > 1) || len > maxCount || len < minCount) {
        // over count limit
        const logStr = len < minCount ? 'under MIN' : 'over MAX';
        const logNum = len < minCount ? minCount : maxCount !== Infinity ? maxCount : 1;
        logger.logWarn(`Upload - The upload total count(${len}) ${logStr} limit(${logNum})`);
        onException && onException('count');
        this._clearInput();
        return false;
      };

      for (let i = 0; i < len; i++) {
        if (!this._handleAccept(files[i].type)) {
          // file type is invalid
          logger.logWarn(`Upload - The ${files[i].name} type is ${files[i].type} which unmatch the accept upload type`);
          onException && onException('accept');
          this._clearInput();
          return false;
        };

        if (files[i].size > size) {
          // file size exceed the limit
          logger.logWarn(`Upload - The ${files[i].name} size(${files[i].size / (1024 * 1024)} MB) exceed the limit(${size / (1024 * 1024)} MB)`);
          onException && onException('size');
          this._clearInput();
          return false;
        };

        changeData.push({
          label: `${getTs()}-${i}`,
          file: files[i]
        });
      };

      onChange && onChange(changeData, e);

      this.fileData = [...this.fileData, ...changeData];
      this._handleRequest(e);
    } else {
      // without file upload
      logger.logErr('Upload - Without any files upload');
    }
    return this._clearInput(); // clear upload object
  }

  public render() {
    const { isMounted, isIE } = this.state;
    const { className, style, prefixCls, disabled } = getProps(this.props);
    const accept = this.getAccept();

    const domProps: DomProps[] = ['id', 'style', 'multiple', 'disabled'];
    const props = pick(this.props, domProps);

    const classes = classnames({
      [prefixCls]: true,
      [`${prefixCls}-adapt-ie`]: isIE,
      [className || '']: !!className
    });

    return (
      <React.Fragment>
        <input
          {...props}
          className={classes}
          type='file'
          accept={accept}
          disabled={isMounted ? disabled : true}
          style={style}
          ref={this.file}
          onChange={this.handleChange}
        />

        { prefixCls === 'hfdc-upload' && <style>
          {`
            .hfdc-upload {
              position: absolute;
              top: 0;
              left: 0;
              width: 100%;
              height: 100%;
              opacity: 0;
              font-size: 0;
              cursor: ${ isMounted ? disabled ? 'not-allowed' : 'pointer' : 'not-allowed'};
            }
            .hfdc-upload-adapt-ie {
              z-index: 99;
              font-size: 100px;
            }
          `}
        </style> }

      </React.Fragment>
    );
  };
};

export default Upload;
