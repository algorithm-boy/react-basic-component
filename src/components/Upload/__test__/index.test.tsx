import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Upload from '../index';

configure({ adapter: new Adapter() });

const mockFile = new File(['foo'], 'foo.png', {
  type: 'image/png'
});

describe('Upload', () => {
  it('renders correctly - shallow', () => {
    const wrapper = shallow(
      <Upload action='http://all-backend.hupu.com/admin/upload' />
    );
    expect(wrapper.find('input').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - render', () => {
    const wrapper = render(
      <button>
        上传按钮
        <Upload action='http://all-backend.hupu.com/admin/upload' />
      </button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(
      <button>
        上传按钮
        <Upload
          action='http://all-backend.hupu.com/admin/upload'
        />
      </button>
    );
    expect(wrapper.contains(<Upload
      action='http://all-backend.hupu.com/admin/upload'
    />)).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it('pass disable parameter', () => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        disabled
      />
    </div>);

    const uploadDOM = wrapper.find('input');
    expect(uploadDOM.props().disabled).toEqual(true);
    uploadDOM.simulate('change');
  });

  it('pass id parameter', () => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        id='upload'
        action='http://all-backend.hupu.com/admin/upload'
      />
    </div>);

    expect(wrapper.find('input').props().id).toEqual('upload');
  });

  it('pass className parameter', () => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        className='hp-upload'
        action='http://all-backend.hupu.com/admin/upload'
      />
    </div>);

    expect(wrapper.find('input').props().className).toEqual('hfdc-upload hp-upload');
  });

  it('pass style parameter', () => {
    const style = {
      width: '300px'
    }

    const wrapper = mount(<div>
      点我上传
      <Upload
        style={style}
        action='http://all-backend.hupu.com/admin/upload'
      />
    </div>);

    expect(wrapper.find('input').props().style).toEqual(style);
  });

  it('pass multiple parameter', () => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        multiple
      />
    </div>);

    const uploadDOM = wrapper.find('input');
    expect(uploadDOM.props().multiple).toEqual(true);
  });

  it('beforeUpload receive params check', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        onChange={files => {
          expect(files).toBeTruthy();
        }}
        beforeUpload={(isRequesting, files, e) => {
          expect(isRequesting).toBeFalsy();
          expect(files).toHaveLength(1);
          expect(e).toBeTruthy();
          done();
          return false
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile]
      }
    });
  }, 10000);

  it('trigger onException - multiple', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        onException={(type) => {
          expect(type).toEqual('count');
          done();
        }}
      />
    </div>);

    const uploadDOM = wrapper.find('input');
    expect(uploadDOM.props().multiple).toEqual(false);
    uploadDOM.simulate('change', {
      target: {
        files: [mockFile, mockFile]
      }
    });
  });

  it('trigger onException - size', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://upload.hupu.com'
        size={-1}
        onException={(type) => {
          expect(type).toEqual('size');
          done();
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile]
      }
    });
  });

  it('trigger onException - count', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        maxCount={1}
        onException={(type) => {
          expect(type).toEqual('count');
          done();
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile, mockFile]
      }
    });
  });

  it('trigger onException - accept', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        accept={'audio'}
        onException={(type) => {
          expect(type).toEqual('accept');
          done();
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile]
      }
    });
  });

  it('trigger onException - api', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://upload.hupu.com'
        onException={(type) => {
          expect(type).toEqual('api');
          done();
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile]
      }
    });
  });

  it('trigger onException - minCount is 3, but upload 2', done => {
    const wrapper = mount(<div>
      点我上传
      <Upload
        action='http://all-backend.hupu.com/admin/upload'
        multiple
        minCount={3}
        onException={(type) => {
          expect(type).toEqual('count');
          done();
        }}
      />
    </div>);

    wrapper.find('input').simulate('change', {
      target: {
        files: [mockFile, mockFile]
      }
    });
  });
});
