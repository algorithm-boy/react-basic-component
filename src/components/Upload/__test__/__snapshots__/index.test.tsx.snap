// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Upload renders correctly - render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "上传按钮",
        "next": Object {
          "attribs": Object {
            "accept": "",
            "class": "hfdc-upload",
            "disabled": "",
            "type": "file",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "
            .hfdc-upload {
              position: absolute;
              top: 0;
              left: 0;
              width: 100%;
              height: 100%;
              opacity: 0;
              font-size: 0;
              cursor: not-allowed;
            }
            .hfdc-upload-adapt-ie {
              z-index: 99;
              font-size: 100px;
            }
          ",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "style",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "style",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "accept": undefined,
            "class": undefined,
            "disabled": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "accept": undefined,
            "class": undefined,
            "disabled": undefined,
            "type": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Object {
        "attribs": Object {
          "accept": "",
          "class": "hfdc-upload",
          "disabled": "",
          "type": "file",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "
            .hfdc-upload {
              position: absolute;
              top: 0;
              left: 0;
              width: 100%;
              height: 100%;
              opacity: 0;
              font-size: 0;
              cursor: not-allowed;
            }
            .hfdc-upload-adapt-ie {
              z-index: 99;
              font-size: 100px;
            }
          ",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "style",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "style",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": Object {
          "data": "上传按钮",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "accept": undefined,
          "class": undefined,
          "disabled": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "accept": undefined,
          "class": undefined,
          "disabled": undefined,
          "type": undefined,
        },
      },
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "
            .hfdc-upload {
              position: absolute;
              top: 0;
              left: 0;
              width: 100%;
              height: 100%;
              opacity: 0;
              font-size: 0;
              cursor: not-allowed;
            }
            .hfdc-upload-adapt-ie {
              z-index: 99;
              font-size: 100px;
            }
          ",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "style",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "accept": "",
            "class": "hfdc-upload",
            "disabled": "",
            "type": "file",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "data": "上传按钮",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "accept": undefined,
            "class": undefined,
            "disabled": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "accept": undefined,
            "class": undefined,
            "disabled": undefined,
            "type": undefined,
          },
        },
        "type": "style",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Upload renders correctly - shallow 1`] = `ShallowWrapper {}`;
